### PR TITLE
Animator event

### PR DESCRIPTION
**Pull request by luzhuang**
### Please check if the PR fulfills these requirements

- [X] The commit message follows our [guidelines](https://github.com/oasis-engine/engine/blob/main/.github/COMMIT_MESSAGE_CONVENTION.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

### Other information:

### CODE REVIEW

File: packages/core/src/animation/Animator.ts [hunk 1]
```diff
@@ -223,24 +227,48 @@ export class Animator extends Component {
...
+        const handler = scripts[j][funcName];
+        if (handler) {
+          const eventHandler = new AnimationEventHandler();
+          eventHandler.event = event;
+          eventHandler.handlers.push(handler);
+          eventHandlers.push(eventHandler);
+        }
+      }
     }
+    eventHandlers.sort((a, b) => b.event.time - a.event.time);
```
Comment (GuoLei1990) [inline packages/core/src/animation/Animator.ts:271]: Sort here is not good! You can sort `AnimatorEvent`  when it is added to `Animationclip`!

File: packages/core/src/animation/Animator.ts [hunk 2]
```diff
@@ -223,24 +227,48 @@ export class Animator extends Component {
...
+    scripts.length = 0;
+    eventHandlers.length = 0;
+    this._entity.getComponents(Script, scripts);
+    for (let i = events.length - 1; i >= 0; i--) {
+      const event = events[i];
+      const funcName = event.functionName;
+      for (let j = scripts.length - 1; j >= 0; j--) {
+        const handler = scripts[j][funcName];
+        if (handler) {
+          const eventHandler = new AnimationEventHandler();
```
Comment (GuoLei1990) [inline packages/core/src/animation/Animator.ts:264]: Why not construct `AnimationEventHandler` in first level loop?
```typescript
for(event)
{
    const eventHandler = new AnimationEventHandler();
     for ( scripts) {
      eventHandler.push(handler);
     }
}
```

File: packages/core/src/animation/Animator.ts [hunk 3]
```diff
@@ -618,4 +650,21 @@ export class Animator extends Component {
...
+      if (time < lastClipTime) continue;
+
+      const { handlers } = eventHandler;
+      if (time === MathUtil.clamp(time, lastClipTime, clipTime)) {
+        for (let j = handlers.length - 1; j >= 0; j--) {
+          handlers[j](parameter);
+        }
+      }
+    }
+  }
```
Comment (GuoLei1990) [inline packages/core/src/animation/Animator.ts:669]: You can do some simple optimizations, for example store current event index to reduce loop count.

**Review by GuoLei1990 — CHANGES_REQUESTED** [review]

File: packages/core/src/animation/Animator.ts (hunk 2 above)
Comment (luzhuang) [reply packages/core/src/animation/Animator.ts:264]: I think we need judge whether the handler existed ,avoid waste

File: packages/core/src/animation/Animator.ts (hunk 3 above)
Comment (luzhuang) [reply packages/core/src/animation/Animator.ts:669]: done

File: packages/core/src/animation/Animator.ts (hunk 1 above)
Comment (luzhuang) [reply packages/core/src/animation/Animator.ts:271]: done

File: packages/core/src/animation/Animator.ts [hunk 4]
```diff
@@ -223,24 +227,48 @@ export class Animator extends Component {
...
+        if (handler) {
+          const eventHandler = new AnimationEventHandler();
+          eventHandler.event = event;
+          eventHandler.handlers.push(handler);
+          eventHandlers.push(eventHandler);
+        }
+      }
     }
+    eventHandlers.sort((a, b) => b.event.time - a.event.time);
   }
```
[the remark this answers is not in the document]
Comment (luzhuang) [reply packages/core/src/animation/Animator.ts:272]: done

File: packages/core/src/animation/Animator.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (luzhuang) [reply packages/core/src/animation/Animator.ts:264]: done

Comment (luzhuang) [reply packages/core/src/animation/Animator.ts:264]: done

File: packages/core/src/animation/Animator.ts [hunk 5]
```diff
@@ -618,4 +653,31 @@ export class Animator extends Component {
...
     }
   }
+
+  private _fireAnimationEvents(eventHandlers: AnimationEventHandler[], lastClipTime: number, clipTime: number) {
+    // TODO: If play backward, not work.
+    let currentIndex = this._currentEventIndex;
+    const len = eventHandlers.length;
+    const firstEventTime = eventHandlers[0].event.time;
+    const lastEventTime = eventHandlers[len - 1].event.time;
+    if (lastEventTime < lastClipTime || firstEventTime >= clipTime) return;
```
Comment (GuoLei1990) [inline packages/core/src/animation/Animator.ts:663]: Maybe don't need to add project code, because the clipTime already safe.

**Review by GuoLei1990 — APPROVED** [review]
+1

**Pull request merged**